Fix block-wise dof renumbering for multigrid

fsonner opened the pull request:
Fixes #871 as proposed by @tjhei .

bangerth commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -657,6 +657,15 @@ namespace DoFRenumbering
+  /**
+   * Sort the degrees of freedom by vector block. It does the same thing as the
+   * above function, only that it does this for one single level of a multi-
+   * level discretization. The non-multigrid part of the the DoFHandler is not
```
don't break words here -- doxygen does not preserve linebreaks in creating output.

bangerth commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -657,6 +657,15 @@ namespace DoFRenumbering
+   */
+  template <int dim, int spacedim>
+  void
+  block_wise (DoFHandler<dim,spacedim> &dof_handler, unsigned int level);
```
can you make the level `const`?

fsonner replied:
That's unnecessary for the declaration as far as I know, but I changed it anyway.

bangerth commented on source/dofs/dof_renumbering.cc:
```diff
@@ -971,20 +972,35 @@ namespace DoFRenumbering
+            //we are dealing with mg dofs, skip foreign level cells:
+            if ((start->get_dof_handler().get_tria().locally_owned_subdomain() != numbers::invalid_subdomain_id)
+                && (cell->level_subdomain_id()
+                    != start->get_dof_handler().get_tria().locally_owned_subdomain()))
```
couldn't you express this condition as
```
  if ( ! cell->is_locally_owned() )
```
or
```
  if ( ! cell->is_locally_owned_on_level() )
```
(you'll need to think about which one is relevant here)?

tjhei replied:
good point. I didn't think of that. Yes, ```is_locally_owned_on_level``` is the correct logic.

fsonner commented:
I implemented your requested changes now. I changed `compute_component_wise` as well as the logic when to skip cells is exactly the same there. I also removed the `cell->is_active()` check inside the `compute_..` functions in case `is_level_operation` is `false` – just skipping cells in case the wrong iterator is passed looks like muting a valuable error message to me.

bangerth commented on include/deal.II/dofs/dof_renumbering.h:
```diff
@@ -657,6 +656,14 @@ namespace DoFRenumbering
+   */
+  template <int dim, int spacedim>
+  void
+  block_wise (DoFHandler<dim,spacedim> &dof_handler, const unsigned int level);
```
Correct, it's unnecessary from a language point of view to declare the parameter `const`. We use it as part of the documentation to make it plain and obvious to the reader that this is an input parameter (that is implicitly already clear because the argument is passed by value, but having a convention obviates the need to think about such things).

tamiko commented:
Looks good. Please rebase (and if you want to clean up the commits, squash).

@bangerth Merge?

tjhei commented:
@tamiko I will take care of this PR. I want to add a few tests first. Give me a couple of days.

tamiko commented:
> @tamiko I will take care of this PR. I want to add a few tests
> first. Give me a couple of days.

Sure.

bangerth commented:
I'll gladly let @tjhei have the last word.

tjhei commented:
Just a heads-up: this renumbering is broken in parallel but I have some stuff on my parallel MG branch that would make it work. I have to figure out how to do this cleanly...

tjhei commented:
Addressed by #890
